lang/funcs/funcgen/pkg.go

```diff
@@ -37,9 +49,175 @@ func parsePkg(path, filename, templates string) error {
 	if err != nil {
 		return err
 	}
-	err = parseFuncs(c, path, templates)
+	functions, err := parsePackages(c)
+	if err != nil {
+		return err
+	}
+	err = parseFuncs(c, functions, path, templates)
 	if err != nil {
 		return err
 	}
 	return nil
 }
+
+func parsePackages(c config) (functions, error) {
+	var funcs []function
+	for _, gopackage := range c.Packages {
+		gopkgfuncs, err := gopackage.Parsefuncs()
+		if err != nil {
+			return funcs, err
+		}
+		funcs = append(funcs, gopkgfuncs...)
+	}
+	return funcs, nil
+}
+
+func (obj *gopackage) Parsefuncs() (functions, error) {
+	var funcs []function
+	cmd := exec.Command("go", "doc", obj.Name)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	err := cmd.Run()
+	if err != nil {
+		return funcs, err
+	}
+	return obj.extractFuncs(out.String())
+}
+
+func (obj *gopackage) extractFuncs(doc string) (functions, error) {
+	var funcs []function
+	for _, line := range strings.Split(doc, "\n") {
+		if validSignature.MatchString(line) {
+			f, err := obj.parseFunctionLine(line)
+			if err != nil && err != excludedErr {
+				return funcs, err
+			}
+			if err == nil {
+				funcs = append(funcs, f)
+			}
+		}
+	}
+
+	return funcs, nil
+}
+
+func (obj *gopackage) parseFunctionLine(line string) (function, error) {
+
+	match := validSignature.FindStringSubmatch(line)
+	result := make(map[string]string)
+	for i, name := range validSignature.SubexpNames() {
+		if i != 0 && name != "" {
+			result[name] = match[i]
+		}
+	}
+
+	name := result["name"]
+
+	for _, e := range obj.Exclude {
+		if e == name {
+			return function{}, excludedErr
+		}
+	}
+
+	errorFul, err := regexp.MatchString(`, error\)$`, result["return"])
+	if err != nil {
+		return function{}, err
+	}
+
+	returns := parseReturn(result["return"])
+	if len(returns) == 0 {
+		return function{}, excludedErr
+	}
+
+	mgmtpackage := obj.Name
+	if obj.MgmtAlias != "" {
+		mgmtpackage = obj.MgmtAlias
+	}
+
+	internalName := fmt.Sprintf("%s_%s", strcase.ToCamel(strings.ReplaceAll(obj.Name, "/", "_")), name)
+	help, err := obj.getHelp(name, internalName)
+	if err != nil {
+		return function{}, err
+	}
+
+	return function{
```

Comment:
please switch all the structs to *function instead.

Reply:
done